test/jdk/com/sun/java/swing/plaf/gtk/TestFileChooserSingleDirectorySelection.java

```diff
@@ -0,0 +1,232 @@
+import java.awt.event.InputEvent;
+import java.awt.Point;
+import java.awt.Robot;
+import javax.swing.JButton;
+import javax.swing.JFileChooser;
+import javax.swing.JFrame;
+import javax.swing.SwingUtilities;
+import javax.swing.UIManager;
+import javax.swing.UnsupportedLookAndFeelException;
+
+public class TestFileChooserSingleDirectorySelection {
+    private static JFrame frame;
+    private static JFileChooser fileChooser;
+    private static JButton getSelectedFilesButton;
+    private static Robot robot;
+    private static boolean passed;
+    private static File[] testDir;
+    private static File[] tempFile;
+
+    public static void main(String[] args) throws Exception {
+        System.setProperty("sun.java2d.uiScale", "1.0");
+        robot = new Robot();
+        robot.setAutoDelay(100);
+
+        try {
+            // create test directory
+            String tmpDir = System.getProperty("user.home");
+            testDir = new File[1];
+            testDir[0] = new File(tmpDir, "testDir");
+            if (!testDir[0].exists())
+                testDir[0].mkdir();
+            testDir[0].deleteOnExit();
+
+            // create temporary files inside testDir
+            tempFile = new File[5];
+            for (int i = 0; i < 5; ++i) {
+                tempFile[i] = File.createTempFile("temp", ".txt",
+                        new File(testDir[0].getAbsolutePath()));
+                tempFile[i].deleteOnExit();
+            }
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+        for (UIManager.LookAndFeelInfo laf :
+                        UIManager.getInstalledLookAndFeels()) {
+            System.out.println("Testing LAF: " + laf.getClassName());
+            SwingUtilities.invokeAndWait(() -> setLookAndFeel(laf));
+            checkFileOnlyTest(laf);
+            checkDirectoriesOnlyTest(laf);
+            checkFilesAndDirectoriesTest(laf);
+            System.out.println("Passed");
+        }
+    }
+
+    private static void checkFileOnlyTest(UIManager.LookAndFeelInfo laf)
+            throws Exception {
+        System.out.println("Testing File Only mode");
+        try {
+            SwingUtilities.invokeAndWait(() -> {
+                createAndShowUI();
+                fileChooser.setCurrentDirectory(testDir[0]);
+            });
+
+            robot.waitForIdle();
+            robot.delay(1000);
+            fileChooser.setFileSelectionMode(JFileChooser.FILES_ONLY);
+            Point frameLocation = fileChooser.getLocationOnScreen();
+            int frameWidth = frame.getWidth();
+            int frameHeight = frame.getHeight();
+
+            Point btnLocation = getSelectedFilesButton.getLocationOnScreen();
+            int btnWidth = getSelectedFilesButton.getWidth();
+            int btnHeight = getSelectedFilesButton.getHeight();
+            clickMouse(frameLocation, 0, frameHeight, 230);
+            clickMouse(btnLocation, btnWidth, btnHeight, 0);
+            checkResult(laf);
```

Comment:
I guess you can optimise by a helper method for this repeated code passing the offset value as paramter.

Also, did you check [JDK-4912623](https://bugs.openjdk.org/browse/JDK-4912623) as mentioned?

Reply:
Added a helper method for the repeated code.

I tried to reproduce the [JDK-4912623](https://bugs.openjdk.org/browse/JDK-4912623) bug. SwingSet2 JFileChooser demo behaves as it is mentioned in the bug description but in Gedit I didn't find separate folder and files list.

Reply:
The bug says 
2) Go to FileChooser Demo.
3) Tab to Folder List
4) Press CTRL+A. It will select all the items in the list.
5) Now Tab to file list and press ctrl+A, it will not select all the items.
6) Now Open Gedit.
7) Click Open.
8) Repeat the steps 3 to 5. Behaviour for the native will be opposite to its Java counterpart.

so you can create manually folder with only files and a folder with only folders and then try with Gedit to open those folders and trey the steps...

Reply:
> The bug says 2) Go to FileChooser Demo. 3) Tab to Folder List 4) Press CTRL+A. It will select all the items in the list. 5) Now Tab to file list and press ctrl+A, it will not select all the items. 6) Now Open Gedit. 7) Click Open. 8) Repeat the steps 3 to 5. Behaviour for the native will be opposite to its Java counterpart.
> 
> so you can create manually folder with only files and a folder with only folders and then try with Gedit to open those folders and trey the steps...

I tried these steps, I am able to select all directories and all files in respective folder on press of Ctrl+A in Gedit.

Reply:
Is GTK Filechooser not behaving the same as gedit?
ie is GTK FileChooser not allowing to select all folders in folder list and all files in files list by CTRL+A with your fix too?

Reply:
With the current fix, GTK FileChooser allows to select all folders in folder list but in files list it allows only single file selection.

Reply:
So, it is same without the fix too as is mentioned in that bug. Can you please see why that is so as it should be exercising the same code path as this fix?

Reply:
> So, it is same without the fix too as is mentioned in that bug. 

Yeah, it behaves same without the fix also.

Reply:
> Can you please see why that is so as it should be exercising the same code path as this fix?

Ok, I will check.

Reply:
> > Can you please see why that is so as it should be exercising the same code path as this fix?
>
> GTK FileChooser allows to select all folders in folder list but in files list it allows only single file selection. The reason behind this is while creating the GTKFileChooserUI, filelist selection model is set to `SINGLE_SELECTION` if **multiselection is disabled**. Code snippet is attached from `protected JScrollPane createFilesList()` method.
       
            if (getFileChooser().isMultiSelectionEnabled()) {
                      fileList.setSelectionMode(ListSelectionModel.MULTIPLE_INTERVAL_SELECTION);
            } else {
                      fileList.setSelectionMode(ListSelectionModel.SINGLE_SELECTION);
            }
But for directory list there is no such condition is present and due to that it allows to select multiple folders in directory list with ctrl+A.

Reply:
@prsadhuk As discussed I have checked with the condition modified to `chooser.getFileSelectionMode() != JFileChooser.FILES_ONLY`, but this leads to failure for current test in `FILES_AND_DIRECTORIES` and `DIRECTORIES_ONLY` selection mode.

And for [JDK-4912623](https://bugs.openjdk.org/browse/JDK-4912623), still not able to select all files in JFileChooser on press of ctrl+A.

Reply:
ok